File: 234. Palindrome Linked List/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        def reverseLinkedList(node):
            current=node
            prev=None
            next=None
            while(current is not None):
                next=current.next
                current.next=prev
                prev=current
                current=next
            return prev
        
        if(head is None or head.next is None):
            return True
        
        slow=head
        fast=head
        while(fast is not None and fast.next is not None):
            fast=fast.next.next
            slow=slow.next
                    
        secondHalf=reverseLinkedList(slow)

        while(secondHalf is not None):
            if(head.val!=secondHalf.val):
                return False
            head=head.next
            secondHalf=secondHalf.next
        return True
        

File: 234. Palindrome Linked List/234. Palindrome Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    var reverseLinkedList=function(node){
        let current=node
        let prev=null
        let next=null
        while(current!==null){
            next=current.next
            current.next=prev
            prev=current
            current=next
        }
        return prev
    }
    if(head===null || head.next===null){
        return true
    }
    var slow=head
    var fast=head
    while(fast!==null && fast.next!==null){
        slow=slow.next
        fast=fast.next.next
    }
    var secondHalf=reverseLinkedList(slow)
    
    while(secondHalf!==null && head!==null){
        if(head.val!==secondHalf.val){
            return false
        }
        head=head.next
        secondHalf=secondHalf.next
    }
    return true   
};